### PR TITLE
Improve `pkgs/by-name` code, minor fix

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

Mainly code cleanup, but that also discovered a problem, which is fixed with the cleanup.

- https://github.com/NixOS/nixpkgs/pull/275539 had false negative bug that would've made https://github.com/NixOS/nixpkgs/pull/284591 succeed CI, even though `sbcl` isn't using `pkgs.callPackage ...` directly.
- CI however failed due to a different false positive bug, which was fixed with https://github.com/NixOS/nixpkgs/pull/285089, though that wasn't implemented fully, only triggering for new packages, but at that point `sbcl` was already merged as is, since I discounted the CI failure as a false positive
- Only now that I cleaned up the code I'm realising the above. This PR fixes it.

See the commit messages for more details.

I'll merge this myself soon after CI is successful.

This work is sponsored by [Antithesis](https://antithesis.com/) and [Tweag](https://www.tweag.io/) :sparkles:

## Things done

- [x] Ran `nix-build -A tests.nixpkgs-check-by-name.tests` successfully on `x86_64-linux`

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by infinisil** [comment]
We can't avoid `pkgs/by-name` CI failing here because it's a bit too clever: It detects that I'm moving `sbcl` out of `pkgs/by-name` again, but this is a rare case where it's totally justified :)

**Comment by hraban** [comment]
I'm not exactly clear on the details but I believe it :) thanks

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs [hunk 1]
```diff
@@ -265,113 +358,117 @@ fn handle_non_by_name_attribute(
...
+    // - Know that the attribute _could_ be migrated to `pkgs/by-name`, which is `Loose`
+    // - Or we're unsure, in which case we use NonApplicable
+    let uses_by_name =
+        // This is a big ol' match on various properties of the attribute
+
+        // First, it needs to succeed evaluation. We can't know whether an attribute could be
+        // migrated to `pkgs/by-name` if it doesn't evaluate, since we need to check that it's a
+        // derivation.
+        //
+        // This only has the minor negative effect that if a PR that breaks evaulation
```
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:374]: evaulation => evaluation

**Review by philiptaron — APPROVED** [review]
I particularly like the changes in `eval.rs`

**Comment by infinisil** [comment]
@hraban The intuitive explanation is that for any `pkgs/by-name/fo/foo`, the corresponding attribute `pkgs.foo` should ideally be defined as `callPackage pkgs/by-name/fo/foo/package.nix { }`. This means that there's a one-to-one mapping from attributes to files without any other necessary context:
- `pkgs.foo` is defined in `pkgs/by-name/fo/foo/package.nix`
- `pkgs/by/name/fo/foo/package.nix` defines `pkgs.foo`

In reality we still need to allow customising the `{ }` argument in `all-packages.nix`. In order to keep the necessary context minimal, CI checks that all definitions for packages in `pkgs/by-name` look like `callPackage pkgs/by-name/fo/foo <attrs>`, where only `<attrs>` can change.

**Comment by infinisil** [comment]
@philiptaron Thanks for the quick review! :heart: 

Typos now fixed :D

**Review by philiptaron — APPROVED** [review]

**Pull request merged**